standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+
+Different parent images should be created in line with each framework's best practice.  Eg, Node.js, .Net Core, Ruby etc.
+
+Parent Dockerfiles should be hosted in GitHub and be deployed to a container registry via CI/CD.
+
+Example Node.js parent image:
+```
+ARG NODE_VERSION=10.18.0
+FROM node:$NODE_VERSION-alpine
+ARG NODE_VERSION
+ARG VERSION=1.0.0
+
+# We need a basic init process to handle signals and reap zombie processes, tini handles that
+RUN apk update && apk add --no-cache tini=0.18.0-r0
+ENTRYPOINT ["/sbin/tini", "--"]
+
+# Never run as root, use the node user from the alpine image
+USER node
+# Default workdir for all the containers that use this image
+WORKDIR /home/node
+
+ENV NODE_ENV production
+# Set global npm dependencies to be stored under the node user directory
+ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+ENV PATH=$PATH:/home/node/.npm-global/bin
+LABEL uk.gov.defra.ffc-node-base.version=$VERSION \
+      uk.gov.defra.ffc-node-base.node-version=$NODE_VERSION
+```
+
+#### Node.js
+- extend a minimal Node base image - e.g. node:alpine
+- set a fixed major version of the base image
+- set the `NODE_ENV` environment variable to production in parent
+- set global NPM dependencies to be stored under the node user directory:
+  ```
+  ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+  ENV PATH=$PATH:/home/node/.npm-global/bin
+  ```
+- run as the `node` user and not `root`
+- ensure the working directory is owned by the node user - e.g. /home/node
+- make the exposed port configurable via a build argument and set as an environment variable:
+  ```
+  ARG PORT=3000
+  ENV PORT ${PORT}
+  ```
+
+#### .Net Core
+- a parent image should exist for both the .Net Core SDK and the .Net Core Runtime
+- extend minimal .Net Core base image
+- set a fixed major version of the base image
+- set the `ASPNETCORE_ENVIRONMENT` environment variable to production in parent
+- run as the `www-data` user and not `root`
+- ensure the working directory is owned by the www-data user - e.g. /app
+- a remote debugger should be added to the SDK file to aid development and testing
+  ```
+  RUN apt-get update \
+  && apt-get install -y --no-install-recommends unzip \
+  && curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg
+  ```
+
+## Container scanning
+Container scanning tools collate vulnerability reports from a number of sources to determine security issues within operating system kernels, libraries, and software packages.
+
+Containers should be regularly scanned using tools such as Clair, Anchore Engine, Aqua Trivy or JFrog XRay.
+
+A comparison of the different tools is available in [Confluence](https://eaflood.atlassian.net/wiki/spaces/FPS/pages/1408172407/Container+Image+Scanning+Tools)
```

Comment:
This sounds like more of an operational, container registry concern than a purely software development one.

I would argue (and have!) that we should be monitoring our _source_ for vulnerabilities and have a CI pipeline that allows us to quickly patch them.  This equally applies if we are maintaining our own container registry, with parent images, as we are building those ourselves.

Being able to scan images that are currently deployed is a secondary concern.  It's still an important thing, but I think it sits more on the operational monitoring side of things.

Reply:
I could shorten to say that we should monitor containers in production without being overly prescriptive?

Reply:
I've removed this section and instead added a line to general saying simply as a standard we will scan running containers for vulnerabilities.